app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/LocalFileTransferActivity.kt

```diff
@@ -0,0 +1,382 @@
+import kotlinx.android.synthetic.main.activity_local_file_transfer.list_peer_devices
+import kotlinx.android.synthetic.main.activity_local_file_transfer.progress_bar_searching_peers
+import kotlinx.android.synthetic.main.activity_local_file_transfer.recycler_view_transfer_files
+import kotlinx.android.synthetic.main.activity_local_file_transfer.text_view_device_name
+import kotlinx.android.synthetic.main.activity_local_file_transfer.text_view_empty_peer_list
+import org.kiwix.kiwixmobile.R
+import org.kiwix.kiwixmobile.core.base.BaseActivity
+import org.kiwix.kiwixmobile.core.di.components.CoreComponent
+import org.kiwix.kiwixmobile.core.extensions.toast
+import org.kiwix.kiwixmobile.core.utils.dialog.AlertDialogShower
+import org.kiwix.kiwixmobile.core.utils.dialog.KiwixDialog
+import org.kiwix.kiwixmobile.kiwixActivityComponent
+import org.kiwix.kiwixmobile.localFileTransfer.WifiDirectManager.Companion.getDeviceStatus
+import org.kiwix.kiwixmobile.localFileTransfer.adapter.WifiP2pDelegate
+import org.kiwix.kiwixmobile.localFileTransfer.adapter.WifiPeerListAdapter
+import java.util.ArrayList
+import javax.inject.Inject
+
+/**
+ * Created by @Aditya-Sood as a part of GSoC 2019.
+ *
+ * This activity is the starting point for the module used for sharing zims between devices.
+ *
+ * The module is used for transferring ZIM files from one device to another, from within the
+ * app. Two devices are connected to each other using WiFi Direct, followed by file transfer.
+ *
+ * File transfer involves two phases:
+ * 1) Handshake with the selected peer device, using [PeerGroupHandshakeAsyncTask]
+ * 2) After handshake, starting the files transfer using [SenderDeviceAsyncTask] on the sender
+ * device and [ReceiverDeviceAsyncTask] files receiving device
+ */
+@SuppressLint("GoogleAppIndexingApiWarning", "Registered")
+class LocalFileTransferActivity : BaseActivity(),
+  WifiDirectManager.Callbacks {
+  @Inject
+  lateinit var alertDialogShower: AlertDialogShower
+
+  @Inject
+  lateinit var wifiDirectManager: WifiDirectManager
+
+  @Inject
+  lateinit var locationManager: LocationManager
+
+  private var isFileSender = false // Whether the device is the file sender or not
+  private var fileListAdapter: FileListAdapter? = null
+  private var wifiPeerListAdapter: WifiPeerListAdapter? = null
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_local_file_transfer)
+    /*
+     * Presence of file Uris decides whether the device with the activity open is a sender or receiver:
+     * - On the sender device, this activity is started from the app chooser post selection
+     * of files to share in the Library
+     * - On the receiver device, the activity is started directly from within the 'Get Content'
+     * activity, without any file Uris
+     * */
+    val filesIntent = intent
+    lateinit var filesForTransfer: List<FileItem>
+    val fileUriArrayList: ArrayList<Uri> =
```

Comment:
Caused by: java.lang.IllegalStateException: filesIntent.getParcelabl…xtra(Intent.EXTRA_STREAM) must not be null
        at org.kiwix.kiwixmobile.localFileTransfer.LocalFileTransferActivity.onCreate(LocalFileTransferActivity.kt:100)

change this type to nullable, immediately crashes if you launch this as a receiver.
Do you know how to test this screen?

Reply:
No sir
However, I've made `fileUriArrayList` nullable.

Reply:
Good but manual testing is essential.

This feature is about the sharing of zim files over wifi direct between phones. You can launch this activity as a receiver of files by clicking on the menu item in the top right of the `Library` destination. To launch this activity as a sender you must have a zim file downloaded, go to `Library`, long press on the book, click the share menu item and select `send via kiwix` or some text similar to this